pkg/networkservice/common/mechanisms/vxlan/vni/client.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2020-2021 Cisco and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vni
+
+import (
+	"context"
+	"net"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/vxlan"
+)
+
+type vniClient struct {
+	tunnelIP net.IP
+}
+
+// NewClient - set the SrcIP for the vxlan mechanism
+func NewClient(tunnelIP net.IP) networkservice.NetworkServiceClient {
+	return &vniClient{
+		tunnelIP: tunnelIP,
+	}
+}
+
+func (v *vniClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*networkservice.Connection, error) {
+	for _, m := range request.GetMechanismPreferences() {
```

Comment:
Basically we expect mechanism client to append its mechanism to `Request.MechanismPreferences`.
Such thing with a loop over the existing preferences is used to guard client if someone have already set needed mechanism but haven't fully filled parameters:
https://github.com/networkservicemesh/sdk/blob/d09236b534179e947c06fabc388a96cae03dc9fe/pkg/networkservice/common/mechanisms/kernel/client.go#L48-L60
1. Try to update
2. If there is nothing to update, then add
So the result of this Request should be needed mechanism in preferences with all needed parameters filled.

Reply:
@Bolodya1997 good find! yes, the vni client is always used along with vxlan client (this chain element is more specific to each forwarder) which sets vxlan mechanism preference and then invokes vni client. 
so adding nil check now, is that fine ?

Reply:
Yep, looks great.